main/src/library/MutPriorityQueue.flix

```diff
@@ -0,0 +1,218 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pub enum MutPriorityQueue[a: Type, r: Region] {
+    case MutPriorityQueue(Region[r], Ref[Array[a, r], r], Ref[Int32, r])
+}
+
+mod MutPriorityQueue {
+
+    ///
+    /// Returns a string representation of `pq`.
+    ///
+    pub def toString(pq: MutPriorityQueue[a, r]): String \ r with ToString[a] = region reg {
+        let MutPriorityQueue(_, arr, s) = pq;
+        let sb = StringBuilder.new(reg);
+        StringBuilder.appendString!("MutPriorityQueue {", sb);
+        Array.forEachWithIndex((i, x) -> {
+            if (i < deref s) {
+                if (i == 0) StringBuilder.appendString!("${x}", sb)
+                else        StringBuilder.appendString!(", ${x}", sb)
+            }
+            else ()
+        }, deref arr);
+        StringBuilder.appendString!("}", sb);
+        StringBuilder.toString(sb)
+    }
+
+    ///
+    /// Returns an empty MutPriorityQueue.
+    ///
+    pub def new(rc: Region[r]): MutPriorityQueue[a, r] \ r =
+        MutPriorityQueue(rc, ref Array.new(rc, 8) @ rc, ref 0 @ rc)
+
+    ///
+    /// Returns the number of elements in `pq`.
+    ///
+    pub def size(pq: MutPriorityQueue[a, r]): Int32 \ r =
+        let MutPriorityQueue(_, _, s) = pq;
+        deref s
+
+    ///
+    /// Returns whether `pq` is empty.
+    ///
+    pub def isEmpty(pq: MutPriorityQueue[a, r]): Bool \ r =
+        let MutPriorityQueue(_, _, s) = pq;
+        deref s == 0
+
+    ///
+    /// Optionally returns the top element of `pq`.
+    ///
+    pub def peek(pq: MutPriorityQueue[a, r]): Option[a] \ r =
+        let MutPriorityQueue(_, arr, s) = pq;
+        if (deref s == 0) None
+        else Some(Array.get(0, deref arr))
+
+    ///
+    /// Enqueues an element `x` into a `pq`.
+    ///
+    pub def enqueue(pq: MutPriorityQueue[a, r], x: a): Unit \ r with Order[a] = {
+        let MutPriorityQueue(_, arr, s) = pq;
+        expand(pq);
+        Array.put(x, deref s, deref arr);
+        heapifyUp(deref s, pq);
+        s := deref s + 1
+    }
+
+    ///
+    /// Removes and optionally returns the top element of `pq`.
+    ///
+    pub def dequeue(pq: MutPriorityQueue[a, r]): Option[a] \ r with Order[a] =
+        let MutPriorityQueue(_, arr, s) = pq;
+        if (deref s > 0) {
+            let top = peek(pq);
+            Array.put(Array.get(deref s - 1, deref arr), 0, deref arr);
+            heapifyDown(0, pq);
+            s := deref s - 1;
+            top
+        }
+        else None
+
+    ///
+    /// Enqueues each element in `l` into `pq`.
+    ///
+    pub def enqueueAll(pq: MutPriorityQueue[a, r], m: m[a]): Unit \ r with Order[a], Iterable[m] = region rc {
```

Comment:
Could we use a foreach loop here?

Reply:
I was confused as to why this would be more efficient

Reply:
And what the alternate option might entail